src/modules/job-exec/job-exec.c

```diff
@@ -374,28 +374,31 @@ static void jobinfo_complete (struct jobinfo *job, const struct idset *ranks)
     }
 }
 
-static int drain_active_ranks (struct jobinfo *job, struct idset *active_ranks)
+int jobinfo_drain_ranks (struct jobinfo *job,
+                         const char *ranks,
+                         const char *fmt,
+                         ...)
 {
-    char *ranks = NULL;
+    va_list ap;
     flux_future_t *f = NULL;
-    char reason[64];
+    char reason[256];
     int rc = -1;
 
-    /* message below guaranteed to fit in < 64 characters */
-    (void) snprintf (reason,
-                     sizeof (reason),
-                     "unkillable processes for job %s",
-                     idf58 (job->id));
-
-    if (!(ranks = idset_encode (active_ranks, IDSET_FLAG_RANGE))
-        || !(f = flux_rpc_pack (job->h,
-                                "resource.drain",
-                                0,
-                                0,
-                                "{s:s s:s s:s}",
-                                "targets", ranks,
-                                "reason", reason,
-                                "mode", "update"))) {
+    /* vsnprintf(3) May truncate (unlikely), but that's ok 255 chars should
+     * get the gist across...
+     */
+    va_start (ap, fmt);
+    (void) vsnprintf (reason, sizeof (reason), fmt, ap);
+    va_end (ap);
+
+    if (!(f = flux_rpc_pack (job->h,
+                             "resource.drain",
+                             0,
+                             0,
+                             "{s:s s:s s:s}",
+                             "targets", ranks,
+                             "reason", reason,
+                             "mode", "update"))) {
         flux_log (job->h,
                   LOG_ERR,
                   "Failed to drain broker ranks %s for job %s",
```

Comment:
Suggestion (minor): change `ranks` back to a `struct idset *` since both callers both have to do the encode otherwise, and then since the two possible failures `idset_encode()` and `flux_rpc_pack()` are both unlikely and we are not checking the RPC response, just consolidate both under a `failed to send drain request for broker ranks...` log.  Also, use `flux_log_error()` to capture the errno.

Reply:
What you suggest was how it was originally implemented, but at least in the start timeout case, the caller needs to do the encode anyway in order to create the job exception message, so it seemed wasteful to encode twice in the same call path since `drain_ranks()` doesn't _need_ a `struct idset`... 

If you feel strongly though I can change it back.